File: CS/15_Document/DocumentAndPagePieceDictionaries.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Windows.Forms;
using Spire.Pdf;


namespace DocumentAndPagePieceDictionaries
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            // Create a new instance of PdfDocument
            PdfDocument pdf = new PdfDocument();

            // Specify the input file path
            String input = @"..\..\..\..\..\..\Data\DocumentAndPagePieceDictionaries.pdf";

            // Load the PDF document from disk
            pdf.LoadFromFile(input);

            /**
             * Add piece dictionaries into a document 
             */

            // If the document piece info is null, create it
            if (pdf.DocumentPieceInfo == null)
            {
                pdf.DocumentPieceInfo = new PdfPieceInfo();
            }

            // Add key-value pairs to the document piece info
            pdf.DocumentPieceInfo.AddApplicationData("ice", "E-iceblue-ice");
            pdf.DocumentPieceInfo.AddApplicationData("blue", "E-iceblue-blue");
            pdf.DocumentPieceInfo.AddApplicationData("Blue", "E-iceblue-Blue");
            pdf.DocumentPieceInfo.AddApplicationData("Ice", "E-iceblue-Ice");

            // Remove a value based on its key from the document piece info
            pdf.DocumentPieceInfo.RemoveApplicationData("blue");

            /**
             * Add piece dictionaries into a page
             */

            // If the piece info in the first page is null, create it
            if (pdf.Pages[0].PagePieceInfo == null)
            {
                pdf.Pages[0].PagePieceInfo = new PdfPieceInfo();
            }

            // Add key-value pairs to the piece info of the first page
            pdf.Pages[0].PagePieceInfo.AddApplicationData("ice", "E-iceblue-ice");
            pdf.Pages[0].PagePieceInfo.AddApplicationData("blue", "E-iceblue-blue");
            pdf.Pages[0].PagePieceInfo.AddApplicationData("Blue", "E-iceblue-Blue");
            pdf.Pages[0].PagePieceInfo.AddApplicationData("Ice", "E-iceblue-Ice");

            // Remove a value based on its key from the piece info of the first page
            pdf.Pages[0].PagePieceInfo.RemoveApplicationData("Ice");

            // Get the piece dictionaries from the document and save them to a file named "documentPieceDictionary.txt"
            getDictionary(pdf.DocumentPieceInfo.ApplicationDatas, "documentPieceDictionary.txt");

            // Get the piece dictionaries from the first page and save them to a file named "pagePieceDictionary.txt"
            getDictionary(pdf.Pages[0].PagePieceInfo.ApplicationDatas, "pagePieceDictionary.txt");

            // Save the modified PDF document to a new file named "DocumentAndPagePieceDictionaries-result.pdf"
            pdf.SaveToFile("DocumentAndPagePieceDictionaries-result.pdf");

            // Close the document
            pdf.Close();

            //Launch the Pdf file
            FileViewer("DocumentAndPagePieceDictionaries-result.pdf");
            //Launch the .txt files
            FileViewer("documentPieceDictionary.txt");
            FileViewer("pagePieceDictionary.txt");
        }

        private void getDictionary(IDictionary<string, PdfApplicationData> dic, string fileName)
        {
            StringBuilder sb = new StringBuilder();

            // Iterate over all the keys in the dictionary
            foreach (string item in dic.Keys)
            {
                PdfApplicationData data = dic[item];
                if (data.Private is String)
                {
                    // Get the value and append it to the StringBuilder
                    string ss = data.Private as string;
                    sb.AppendLine(ss);
                }
            }

            // Write the text content of the StringBuilder to the specified file
            File.WriteAllText(fileName, sb.ToString());
        }

        private void FileViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }

    }
}

File: VB/15_Document/DocumentAndPagePieceDictionaries.vb
Imports System.IO
Imports System.Text
Imports Spire.Pdf

Namespace DocumentAndPagePieceDictionaries
    Partial Public Class Form1
        Inherits Form
        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
            ' Create a new PdfDocument object
            Dim pdf As New PdfDocument()

            ' Specify the path of the input PDF file
            Dim input As String = "..\..\..\..\..\..\Data\DocumentAndPagePieceDictionaries.pdf"

            ' Load the PDF document from the specified path
            pdf.LoadFromFile(input)

            ' * Add piece dictionaries into a document  * 

            ' Check if the document's piece information is null
            If pdf.DocumentPieceInfo Is Nothing Then
                ' If it is null, create a new PdfPieceInfo object and assign it to the document's PieceInfo property
                pdf.DocumentPieceInfo = New PdfPieceInfo()
            End If

            ' Add application data to the document's piece information
            pdf.DocumentPieceInfo.AddApplicationData("ice", "E-iceblue-ice")
            pdf.DocumentPieceInfo.AddApplicationData("blue", "E-iceblue-blue")
            pdf.DocumentPieceInfo.AddApplicationData("Blue", "E-iceblue-Blue")
            pdf.DocumentPieceInfo.AddApplicationData("Ice", "E-iceblue-Ice")

            ' Remove the application data with key "blue" from the document's piece information
            pdf.DocumentPieceInfo.RemoveApplicationData("blue")


            ' * Add piece dictionaries into a page * 

            ' Check if the page's piece information is null
            If pdf.Pages(0).PagePieceInfo Is Nothing Then
                ' If it is null, create a new PdfPieceInfo object and assign it to the page's PieceInfo property
                pdf.Pages(0).PagePieceInfo = New PdfPieceInfo()
            End If

            ' Add application data to the page's piece information
            pdf.Pages(0).PagePieceInfo.AddApplicationData("ice", "E-iceblue-ice")
            pdf.Pages(0).PagePieceInfo.AddApplicationData("blue", "E-iceblue-blue")
            pdf.Pages(0).PagePieceInfo.AddApplicationData("Blue", "E-iceblue-Blue")
            pdf.Pages(0).PagePieceInfo.AddApplicationData("Ice", "E-iceblue-Ice")

            ' Remove the application data with key "Ice" from the page's piece information
            pdf.Pages(0).PagePieceInfo.RemoveApplicationData("Ice")

            ' Get the dictionary of application datas from the document's piece information and save it to a file named "documentPieceDictionary.txt"
            getDictionary(pdf.DocumentPieceInfo.ApplicationDatas, "documentPieceDictionary.txt")

            ' Get the dictionary of application datas from the page's piece information and save it to a file named "pagePieceDictionary.txt"
            getDictionary(pdf.Pages(0).PagePieceInfo.ApplicationDatas, "pagePieceDictionary.txt")

            ' Save the modified PDF document to a file named "DocumentAndPagePieceDictionaries-result.pdf"
            pdf.SaveToFile("DocumentAndPagePieceDictionaries-result.pdf")

            ' Close the PDF document
            pdf.Close()

            'Launch the Pdf file
            FileViewer("DocumentAndPagePieceDictionaries-result.pdf")

            'Launch the .txt files
            FileViewer("documentPieceDictionary.txt")
            FileViewer("pagePieceDictionary.txt")
        End Sub

        Private Sub getDictionary(ByVal dic As IDictionary(Of String, PdfApplicationData), ByVal fileName As String)
            Dim sb As New StringBuilder()
            'Traverse all keys in the dictionary
            For Each item As String In dic.Keys
                Dim data As PdfApplicationData = dic(item)
                If TypeOf data.Private Is String Then
                    'Get the value and append it to StringBuilder
                    Dim ss As String = TryCast(data.Private, String)
                    sb.AppendLine(ss)
                End If
            Next item
            'Wirte the text of StringBuilder to file
            File.WriteAllText(fileName, sb.ToString())
        End Sub

        Private Sub FileViewer(ByVal fileName As String)
            Try
                Process.Start(fileName)
            Catch
            End Try
        End Sub

    End Class
End Namespace
